Use more constrained types to represent CSS gradient values

weinig opened the pull request:
#### f62447bff59b85890116e58a9740fc50113bba06
<pre>
Use more constrained types to represent CSS gradient values
<a href="https://bugs.webkit.org/show_bug.cgi?id=247235">https://bugs.webkit.org/show_bug.cgi?id=247235</a>
rdar://101720047

Reviewed by Darin Adler.

To make further progress on moving the CSS gradient code toward
correct semantics, I took a moment to refactor the current CSS
value subclasses to better represent the parsed output. To do
this, I broke apart the deprecated, prefixed and standard variants
into their own types and now only store exactly what is needed
to represent the gradient, utilizing std::variant and additional
structs to provide the structure.

Now, if a gradient has a member that is a RefPtr&lt;&gt;, you can be
assured that it is required to check for nullptr, otherwise, a
Ref&lt;&gt; would have been used.

I also took the opertunity to remove the base CSSGradientValue
class as there was no need polymorphism being used and removing
it gets us closer to CSSValue value semantics.

* Source/WebCore/css/CSSGradientValue.cpp:
(WebCore::computeStops):
(WebCore::CSSLinearGradientValue::createStyleImage const):
(WebCore::CSSPrefixedLinearGradientValue::createStyleImage const):
(WebCore::CSSDeprecatedLinearGradientValue::createStyleImage const):
(WebCore::CSSRadialGradientValue::createStyleImage const):
(WebCore::CSSPrefixedRadialGradientValue::createStyleImage const):
(WebCore::CSSDeprecatedRadialGradientValue::createStyleImage const):
(WebCore::CSSConicGradientValue::createStyleImage const):
(WebCore::operator==):
(WebCore::cssText):
(WebCore::CSSLinearGradientValue::customCSSText const):
(WebCore::CSSLinearGradientValue::equals const):
(WebCore::CSSPrefixedLinearGradientValue::customCSSText const):
(WebCore::CSSPrefixedLinearGradientValue::equals const):
(WebCore::CSSDeprecatedLinearGradientValue::customCSSText const):
(WebCore::CSSDeprecatedLinearGradientValue::equals const):
(WebCore::CSSRadialGradientValue::customCSSText const):
(WebCore::CSSRadialGradientValue::equals const):
(WebCore::CSSPrefixedRadialGradientValue::customCSSText const):
(WebCore::CSSPrefixedRadialGradientValue::equals const):
(WebCore::CSSDeprecatedRadialGradientValue::customCSSText const):
(WebCore::CSSDeprecatedRadialGradientValue::equals const):
(WebCore::CSSConicGradientValue::customCSSText const):
(WebCore::CSSConicGradientValue::equals const):
(WebCore::CSSGradientValue::computeStops const): Deleted.
(WebCore::CSSGradientValue::createStyleImage const): Deleted.
(WebCore::CSSGradientValue::equals const): Deleted.
* Source/WebCore/css/CSSGradientValue.h:
(WebCore::CSSGradientValue::setFirstX): Deleted.
(WebCore::CSSGradientValue::setFirstY): Deleted.
(WebCore::CSSGradientValue::setSecondX): Deleted.
(WebCore::CSSGradientValue::setSecondY): Deleted.
(WebCore::CSSGradientValue::gradientType const): Deleted.
(WebCore::CSSGradientValue::CSSGradientValue): Deleted.
(WebCore::CSSGradientValue::firstX const): Deleted.
(WebCore::CSSGradientValue::firstY const): Deleted.
(WebCore::CSSGradientValue::secondX const): Deleted.
(WebCore::CSSGradientValue::secondY const): Deleted.
(WebCore::CSSGradientValue::stops const): Deleted.
(WebCore::CSSGradientValue::isRepeating const): Deleted.
(WebCore::CSSGradientValue::colorInterpolationMethod const): Deleted.
* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::visitDerived):
* Source/WebCore/css/CSSValue.h:
(WebCore::CSSValue::isImageGeneratorValue const):
(WebCore::CSSValue::isGradientValue const):
(WebCore::CSSValue::isDeprecatedLinearGradientValue const):
(WebCore::CSSValue::isDeprecatedRadialGradientValue const):
(WebCore::CSSValue::isPrefixedLinearGradientValue const):
(WebCore::CSSValue::isPrefixedRadialGradientValue const):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeDeprecatedGradientPointValue):
(WebCore::CSSPropertyParserHelpers::consumeDeprecatedGradientPoint):
(WebCore::CSSPropertyParserHelpers::consumeDeprecatedGradientColorStops):
(WebCore::CSSPropertyParserHelpers::consumeDeprecatedLinearGradient):
(WebCore::CSSPropertyParserHelpers::consumeDeprecatedRadialGradient):
(WebCore::CSSPropertyParserHelpers::consumeDeprecatedGradient):
(WebCore::CSSPropertyParserHelpers::consumeColorStopList):
(WebCore::CSSPropertyParserHelpers::consumeLengthColorStopList):
(WebCore::CSSPropertyParserHelpers::consumeAngularColorStopList):
(WebCore::CSSPropertyParserHelpers::consumePrefixedRadialGradient):
(WebCore::CSSPropertyParserHelpers::consumeRadialGradient):
(WebCore::CSSPropertyParserHelpers::consumePrefixedLinearGradient):
(WebCore::CSSPropertyParserHelpers::consumeLinearGradient):
(WebCore::CSSPropertyParserHelpers::consumeConicGradient):
(WebCore::CSSPropertyParserHelpers::consumeGeneratedImage):
(WebCore::CSSPropertyParserHelpers::consumeGradientColorStops): Deleted.
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::autoFillStrongPasswordMaskImage):
* Source/WebCore/rendering/style/StyleGradientImage.cpp:
(WebCore::operator==):
(WebCore::StyleGradientImage::StyleGradientImage):
(WebCore::StyleGradientImage::equals const):
(WebCore::StyleGradientImage::computedStyleValue const):
(WebCore::StyleGradientImage::computeStopsForDeprecatedVariants const):
(WebCore::StyleGradientImage::computeStops const):
(WebCore::positionFromValue):
(WebCore::computeEndPoint):
(WebCore::endPointsFromAngle):
(WebCore::endPointsFromAngleForPrefixedVariants):
(WebCore::resolveRadius):
(WebCore::distanceToClosestCorner):
(WebCore::distanceToFarthestCorner):
(WebCore::StyleGradientImage::createGradient const):
* Source/WebCore/rendering/style/StyleGradientImage.h:
* Source/WebCore/style/StyleBuilderState.cpp:
(WebCore::Style::BuilderState::createStyleImage):
* Source/WebCore/style/StylePendingResources.cpp:

Canonical link: <a href="https://commits.webkit.org/256211@main">https://commits.webkit.org/256211@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/adc6def57f52e181923ff089076168bb35e4f1c9

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/95129 "9 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/4269 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/28153 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104726 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/164976 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/99121 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/4363 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/33105 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/87441 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100623 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100797 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/3177 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/81663 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/30158 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/85086 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/86921 "Build was cancelled. Recent messages:Configured build (cancelled)") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/73053 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38859 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/18476 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36679 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/19757 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4295 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40614 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/42452 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/42597 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38999 "Passed tests") | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -3434,94 +3447,126 @@ static bool consumeDeprecatedGradientColorStop(CSSParserTokenRange& range, CSSGr
 }
 
-static RefPtr<CSSValue> consumeDeprecatedGradient(CSSParserTokenRange& args, const CSSParserContext& context)
+static RefPtr<CSSValue> consumeDeprecatedLinearGradient(CSSParserTokenRange& args, const CSSParserContext& context)
```
I am confused about all these consume functions, and I figured you might know the answer.

Which ones are allowed to consume some partial parsed content when they fail? Or is the contract that they must only move the range forward if the succeed?

Also, when should we name this `args` and when should we name it `range`?

Not really a comment in this patch, but wanting to improve our CSS parser idiom clarity.

weinig replied:
These "gradient" ones are all able to consume as much as they want, as the args they are using are a copy made in `consumeGeneratedImage` and if they fail, the whole parse fails anyway.

But, I think the larger question of "how do you know that" from looking at a function is a good one. In this case, maybe the functions should actually take the token range by value (or maybe r-value) to make it super clear. (We would also have to make them check `args.atEnd()`, which is currently the only reason they are being passed by reference.)

Maybe a naming scheme would help to differentiate?

weinig replied:
Also, hahaha that you are reviewing this. I was trying to be silly :). By the way, your PR, https://github.com/WebKit/WebKit/pull/5986, is looking great! I started putting to words some ideas for CSS properties in the new wiki - https://github.com/WebKit/WebKit/wiki/DRAFT:-Proposal-for-Generated-&-Modernized-CSS-Properties-and-Values

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -3559,51 +3604,123 @@ static std::optional<CSSGradientColorStopList> consumeGradientColorStops(CSSPars
+    auto consumeExtentKeyword = [](auto& args) -> std::optional<CSSPrefixedRadialGradientValue::ExtentKeyword> {
+        switch (args.peek().id()) {
+        case CSSValueClosestSide:
+            args.consumeIncludingWhitespace();
```
So many different idioms for consuming expected identifiers! This one is fine.

weinig replied:
Yeah, going to make a template function to fix this. https://bugs.webkit.org/show_bug.cgi?id=247326. I think I can use SortedArrayMap to make it nice.

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
```diff
@@ -3559,51 +3604,123 @@ static std::optional<CSSGradientColorStopList> consumeGradientColorStops(CSSPars
+        gradientBox = *extentKeyword;
+    } else {
+        auto length1 = consumeLengthOrPercent(args, context.mode, ValueRange::NonNegative);
+        if (length1) {
```
Scope this one inside the if statement?

grorg commented on Source/WebCore/html/HTMLInputElement.cpp:
```diff
@@ -2154,10 +2154,11 @@ static Ref<StyleGradientImage> autoFillStrongPasswordMaskImage()
+            {
+                CSSLinearGradientValue::Angle { CSSValuePool::singleton().createValue(90, CSSUnitType::CSS_DEG) }
+            },
+            CSSGradientRepeat::NonRepeating
```
I never knew the `input` element implementation had gradients, but now I do I remember exactly where it is. And that UI sort of annoys me :)

Commit:
Use more constrained types to represent CSS gradient values
https://bugs.webkit.org/show_bug.cgi?id=247235
rdar://101720047

Reviewed by Darin Adler.

To make further progress on moving the CSS gradient code toward
correct semantics, I took a moment to refactor the current CSS
value subclasses to better represent the parsed output. To do
this, I broke apart the deprecated, prefixed and standard variants
into their own types and now only store exactly what is needed
to represent the gradient, utilizing std::variant and additional
structs to provide the structure.

Now, if a gradient has a member that is a RefPtr<>, you can be
assured that it is required to check for nullptr, otherwise, a
Ref<> would have been used.

I also took the opertunity to remove the base CSSGradientValue
class as there was no need polymorphism being used and removing
it gets us closer to CSSValue value semantics.

* Source/WebCore/css/CSSGradientValue.cpp:
(WebCore::computeStops):
(WebCore::CSSLinearGradientValue::createStyleImage const):
(WebCore::CSSPrefixedLinearGradientValue::createStyleImage const):
(WebCore::CSSDeprecatedLinearGradientValue::createStyleImage const):
(WebCore::CSSRadialGradientValue::createStyleImage const):
(WebCore::CSSPrefixedRadialGradientValue::createStyleImage const):
(WebCore::CSSDeprecatedRadialGradientValue::createStyleImage const):
(WebCore::CSSConicGradientValue::createStyleImage const):
(WebCore::operator==):
(WebCore::cssText):
(WebCore::CSSLinearGradientValue::customCSSText const):
(WebCore::CSSLinearGradientValue::equals const):
(WebCore::CSSPrefixedLinearGradientValue::customCSSText const):
(WebCore::CSSPrefixedLinearGradientValue::equals const):
(WebCore::CSSDeprecatedLinearGradientValue::customCSSText const):
(WebCore::CSSDeprecatedLinearGradientValue::equals const):
(WebCore::CSSRadialGradientValue::customCSSText const):
(WebCore::CSSRadialGradientValue::equals const):
(WebCore::CSSPrefixedRadialGradientValue::customCSSText const):
(WebCore::CSSPrefixedRadialGradientValue::equals const):
(WebCore::CSSDeprecatedRadialGradientValue::customCSSText const):
(WebCore::CSSDeprecatedRadialGradientValue::equals const):
(WebCore::CSSConicGradientValue::customCSSText const):
(WebCore::CSSConicGradientValue::equals const):
(WebCore::CSSGradientValue::computeStops const): Deleted.
(WebCore::CSSGradientValue::createStyleImage const): Deleted.
(WebCore::CSSGradientValue::equals const): Deleted.
* Source/WebCore/css/CSSGradientValue.h:
(WebCore::CSSGradientValue::setFirstX): Deleted.
(WebCore::CSSGradientValue::setFirstY): Deleted.
(WebCore::CSSGradientValue::setSecondX): Deleted.
(WebCore::CSSGradientValue::setSecondY): Deleted.
(WebCore::CSSGradientValue::gradientType const): Deleted.
(WebCore::CSSGradientValue::CSSGradientValue): Deleted.
(WebCore::CSSGradientValue::firstX const): Deleted.
(WebCore::CSSGradientValue::firstY const): Deleted.
(WebCore::CSSGradientValue::secondX const): Deleted.
(WebCore::CSSGradientValue::secondY const): Deleted.
(WebCore::CSSGradientValue::stops const): Deleted.
(WebCore::CSSGradientValue::isRepeating const): Deleted.
(WebCore::CSSGradientValue::colorInterpolationMethod const): Deleted.
* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::visitDerived):
* Source/WebCore/css/CSSValue.h:
(WebCore::CSSValue::isImageGeneratorValue const):
(WebCore::CSSValue::isGradientValue const):
(WebCore::CSSValue::isDeprecatedLinearGradientValue const):
(WebCore::CSSValue::isDeprecatedRadialGradientValue const):
(WebCore::CSSValue::isPrefixedLinearGradientValue const):
(WebCore::CSSValue::isPrefixedRadialGradientValue const):
* Source/WebCore/css/parser/CSSPropertyParserHelpers.cpp:
(WebCore::CSSPropertyParserHelpers::consumeDeprecatedGradientPointValue):
(WebCore::CSSPropertyParserHelpers::consumeDeprecatedGradientPoint):
(WebCore::CSSPropertyParserHelpers::consumeDeprecatedGradientColorStops):
(WebCore::CSSPropertyParserHelpers::consumeDeprecatedLinearGradient):
(WebCore::CSSPropertyParserHelpers::consumeDeprecatedRadialGradient):
(WebCore::CSSPropertyParserHelpers::consumeDeprecatedGradient):
(WebCore::CSSPropertyParserHelpers::consumeColorStopList):
(WebCore::CSSPropertyParserHelpers::consumeLengthColorStopList):
(WebCore::CSSPropertyParserHelpers::consumeAngularColorStopList):
(WebCore::CSSPropertyParserHelpers::consumePrefixedRadialGradient):
(WebCore::CSSPropertyParserHelpers::consumeRadialGradient):
(WebCore::CSSPropertyParserHelpers::consumePrefixedLinearGradient):
(WebCore::CSSPropertyParserHelpers::consumeLinearGradient):
(WebCore::CSSPropertyParserHelpers::consumeConicGradient):
(WebCore::CSSPropertyParserHelpers::consumeGeneratedImage):
(WebCore::CSSPropertyParserHelpers::consumeGradientColorStops): Deleted.
* Source/WebCore/html/HTMLInputElement.cpp:
(WebCore::autoFillStrongPasswordMaskImage):
* Source/WebCore/rendering/style/StyleGradientImage.cpp:
(WebCore::operator==):
(WebCore::StyleGradientImage::StyleGradientImage):
(WebCore::StyleGradientImage::equals const):
(WebCore::StyleGradientImage::computedStyleValue const):
(WebCore::StyleGradientImage::computeStopsForDeprecatedVariants const):
(WebCore::StyleGradientImage::computeStops const):
(WebCore::positionFromValue):
(WebCore::computeEndPoint):
(WebCore::endPointsFromAngle):
(WebCore::endPointsFromAngleForPrefixedVariants):
(WebCore::resolveRadius):
(WebCore::distanceToClosestCorner):
(WebCore::distanceToFarthestCorner):
(WebCore::StyleGradientImage::createGradient const):
* Source/WebCore/rendering/style/StyleGradientImage.h:
* Source/WebCore/style/StyleBuilderState.cpp:
(WebCore::Style::BuilderState::createStyleImage):
* Source/WebCore/style/StylePendingResources.cpp:

Canonical link: https://commits.webkit.org/256211@main
